source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -43,101 +56,294 @@ template <class RequestType, class ResponseType> class GrpcMuxFailover {
       GrpcStreamCallbacks<ResponseType>* stream_callbacks)>;
 
   GrpcMuxFailover(GrpcStreamCreator primary_stream_creator,
-                  OptRef<GrpcStreamCreator> failover_stream_creator,
-                  GrpcStreamCallbacks<ResponseType>& grpc_mux_callbacks)
+                  absl::optional<GrpcStreamCreator> failover_stream_creator,
+                  GrpcStreamCallbacks<ResponseType>& grpc_mux_callbacks,
+                  Event::Dispatcher& dispatcher)
       : grpc_mux_callbacks_(grpc_mux_callbacks), primary_callbacks_(*this),
-        primary_grpc_stream_(std::move(primary_stream_creator(&primary_callbacks_))) {
+        primary_grpc_stream_(std::move(primary_stream_creator(&primary_callbacks_))),
+        connecting_to_primary_(false), connecting_to_failover_(false),
+        connected_to_(ConnectedState::None), ever_connected_to_(ConnectedState::None) {
     ASSERT(primary_grpc_stream_ != nullptr);
-    // At the moment failover isn't implemented.
-    ASSERT(!failover_stream_creator.has_value());
+    if (failover_stream_creator.has_value()) {
+      ENVOY_LOG(warn, "Using xDS-Failover. Note that the implementation is currently considered "
+                      "experimental and may be modified in future Envoy versions!");
+      // Only create the retry timer if failover is supported.
+      complete_retry_timer_ = dispatcher.createTimer([this]() -> void { retryConnections(); });
+      failover_callbacks_ = std::make_unique<FailoverGrpcStreamCallbacks>(*this);
+      GrpcStreamCreator& failover_stream_creator_ref = failover_stream_creator.value();
+      failover_grpc_stream_ = std::move(failover_stream_creator_ref(failover_callbacks_.get()));
+      ASSERT(failover_grpc_stream_ != nullptr);
+    }
   }
 
   virtual ~GrpcMuxFailover() = default;
 
+  // Attempts to establish a new stream to the either the primary or failover source.
   void establishNewStream() {
```

Comment:
this method has certain degree of code duplication, wonder if it would be possible to refactor.